### ISSUE TITLE
[4.x]: GQL: Default siteId is different between graphiql and queries from outside

### ISSUE THREAD

**Issue by tremby** [craftcms/cms#10917]
### Description

When querying with GQL in graphiql, it seems to default to giving results as if I'd asked for `siteId: 1`. If I query instead from elsewhere, with a bearer token (which allows access to all my sites), it seems to default to giving results as if I'd asked for `siteId: 2`.

### Steps to reproduce

1. Have a multisite Craft with a couple of different sites
2. Have entries for multiple sites
3. Query for `{ entries { id, siteId } }` in graphiql
4. Query for `{ entries { id, siteId } }` from elsewhere such as curl, with a bearer token which allows access to all sites

### Expected behavior

A consistent default siteId, or maybe for it to be as if I've asked for results from all sites at once.

### Actual behavior

It seems to default to site 1 in graphiql, and site 2 when I query from curl with my bearer token.

### Craft CMS version

4 beta at 792e226

### PHP version

8.1.2

### Operating system and version

Linux 5.13.0

### Database type and version

Mysql 8.0.27

### Installed plugins and versions

- Control Panel CSS v2.x-dev
- Formie dev-craft-4
- Navigation dev-craft-4
- Redactor 3.0.0-beta.2
- Retour 4.0.0-beta.4
- SEOmatic 4.0.0-beta.7+
- Typed link field 2.1.0-rc
- Vizy dev-craft-4
- Workflow dev-craft-4

**Comment by andris-sevcenko** [comment]
If no `siteId` is specified, Craft defaults to the accessed site. I'm guessing that you're targeting your GQL query to a different site than where you're accessing the control panel? If, that's the expected behavior.

**Comment by tremby** [comment]
No, I'm not.

The graphql endpoint is currently below the control panel URL. None of the
sites' base URLs look anything like the control panel's URL. I'm not sure
how Craft would choose either of the two configured sites as being
"current" based on the GQL endpoint, nor why it would change its mind based
on whether I'm querying via graphiql or via curl.

To illustrate, the control panel's URL is similar to http://dev.example.com
and the graphql endpoint is /api on top of that. Right now for dev purposes
the default site's base URL is http://localhost:3000 and the other site's
base URL is http://localhost:3001.

**Comment by andris-sevcenko** [comment]
Is one of the sites disabled? Frontend requests (which `/api` is) don't list disabled sites, if memory serves.

Either way, if no siteId is specified, then no siteId gets passed to the element queries underneath GraphQL, either, so Craft ends up using what it thinks is the current site. 

If all your site base URLs are `http://localhost:*` and you're accessing the API on a different domain, the site selection behaviour would be defined, as Craft would struggle matching any of those Base URLs against the server URL.

**Comment by tremby** [comment]
None of them are disabled. And I probably should have mentioned that Craft is configured to be in headless mode.

**Comment by tremby** [comment]
> Either way, if no siteId is specified, then no siteId gets passed to the element queries underneath GraphQL, either, so Craft ends up using what it thinks is the current site.

Where exactly is that logic which decides the current site? Maybe if I can read the code I'll figure out why it's sometimes choosing one and sometimes choosing another.

> If all your site base URLs are `http://localhost:*` and you're accessing the API on a different domain, the site selection behaviour would be defined, as Craft would struggle matching any of those Base URLs against the server URL.

OK, well that seems like a bug -- undefined behaviour is bad. If it can't conclusively say which site is current, it should either pick the default site or pick *all* sites; either would make sense.

I'm not sure why Craft would expect the API, which is a back-end feature, to be on the same domain as any of the front ends other than the control panel, especially in a headless multisite situation.

**Comment by tremby** [comment]
To answer my first question, the logic to determine the current site is in [`src/web/Request.php`](https://github.com/craftcms/cms/blob/4.0/src/web/Request.php), with various functions helping, fired off from `init`. The logic is somewhat elaborate to my eye (it's 1000 points for a partial string match plus 100 points for each successive character matched, and then +100 if the port matches and +10 if the scheme does, and +1 if it's the default site), and part of me thinks it'll just lead to false positives and hard-to-trace bugs. Then again, I can see the utility of partial matches since it'll help in cases like a stage or dev server with a similar hostname to production, but which is still reading the production configuration. I'd sort of like a way to opt out of it, to opt in to something stricter, but that's a different conversation.

Since I started digging I ran my queries again since I wanted some debugging output, only to find that it's now behaving -- it's defaulting to site 1 in all cases. I think probably what happened is that yesterday something was still cached (GraphQL result, most likely) from before I set the URLs for each site. I've now switched off Graphql caching for development.

It does look like the code is deterministic with given inputs; I think there's no bug here.
